accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+# Link npm packages to the originating source code repository and build
+
+**Authors**: @feelepxyz @kommendorkapten @trevrosen
+
+## Summary
+Link public npm packages to the source code repository and build it originated from, making it harder to execute package hijacking attacks where a malicious version of an existing open source package gets uploaded to the registry.
+
+Check out the [glossary](#glossary) for an overview of the different tools, terms and techniques covered in this RFC document.
+
+## Motivation
+So far our supply chain security offerings have focused on remediation - patching known vulnerabilities in dependencies (e.g. `npm audit`). The majority of vulnerabilities are accidental, so that was the right place to start. Now we need to turn our attention to deliberate supply chain attacks. In 2021, we saw these types of attacks increase by a terrifying [650%](https://www.sonatype.com/resources/state-of-the-software-supply-chain-2021).
+
+In November of last year alone, we saw three supply chain attacks against popular npm packages ([UA-Parser-JS](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack), [COA](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack), and [RC](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack)), each with millions of downloads per month.
+
+The following chart illustrates several attack vectors facing the supply chain:
+
+![supply-chain-threats](https://user-images.githubusercontent.com/20165/174750708-2be483ac-7e41-4bc3-8ee9-440ef33d9423.svg)
+
+Our focus is on mitigating package hijacking attacks where packages are built from a modified source (C) in the above diagram. This is where a malicious version of an existing open source package gets uploaded to the registry. These maliciously crafted versions are rarely reflected in the canonical source code repository and uploaded directly from an attacker's machine.
+
+The [above](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack) [attacks](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack) [are](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack) examples of this, and they frequently occur due to compromised npm credentials but can also happen due to compromised CI/CD or builds.
+
+Code on GitHub, GitLab etc can be browsed and audited, but packages in the registry are opaque and much harder to vet. If packages are built and published in the open these attacks become a whole lot harder. However, right now there’s no way of knowing where a package came from when you retrieve it from the registry.
+
+We want to solve the problem of npm packages being disconnected from their source by linking the published package to the source code repository and build that it originated from.
```

Comment:
given that build output should never be committed, and builds/publishes are typically done on local machines, how can this link be verified?

Reply:
This RFC will only work with packages published from trusted build infrastructure. It will not support local publish.

Reply:
What are the criteria for "trusted"? What are the associated costs in standing up such infrastructure, and what portions of the community would be excluded by those costs?

Reply:
@ljharb I believe the premise of this entire RFC is predicated on builds/publishes happening in the cloud (as the motivation aims at getting closer to being able to meet SLSA Level 3 standards - ref. https://github.com/npm/rfcs/blob/link-packages-to-source-and-build/accepted/0000-link-packages-to-source-and-build.md#non-falsifiable-provenance-using-a-trusted-builder & https://slsa.dev/spec/v0.1/levels#detailed-explanation)

Reply:
This is all covered in the RFC

https://github.com/npm/rfcs/pull/626/files#diff-95858aa7674c6ecb2578ad8c2b7ac848ac80dd36eedbbc92ce8a3968daf10771R239-R249

Reply:
@darcyclarke without staged builds, I don't think that's something that the ecosystem will be able to adopt. Is that part of this RFC?

Reply:
> @darcyclarke without staged builds, I don't think that's something that the ecosystem will be able to adopt. Is that part of this RFC?

Staged builds are not currently part of this RFC but I can see how that could make it easier and safer to adopt automated publishes from ci/cd.

Reply:
@ljharb I think we can use this RFC process to determine if your hypothesis is correct. Personally I do not believe that staged builds as a direct feature of npm will be necessary for ecosystem adoption and is not currently on our product roadmap.

It is possible to implement a ["manual approval" in GitHub actions](https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments) and I've seen customer evidence that [they are able to use this for npm today](https://twitter.com/meijer_s/status/1539535747212009474).

What is on the product roadmap are a number of improvements related to making publishing from CI more secure such as access tokens scoped to specific packages / scopes and streamlined setup experience to make it easier to get started.

Reply:
If the manual approval allows me to input an actual second factor token, then that would be great! The things you linked don't seem to permit that yet tho.

I look forward to the RFC discussion.

Reply:
@ljharb I'm surfacing that request to the actions team.

In general I see staging / approvals as out of scope for npm as a whole, but up for challenging those assumptions.

Reply:
I would like to point out that the one of the major use cases of SLSA provenance is to mitigate the threat of compromised (Npm) registry credentials.  That's because the provenance provides a link to the source repo. So if an attacker were to compromise a registry credential and push a malicious package, the provenance would change from "source=github.com/genuine/source" to "source=github.com/attacker/source" or even "". The registry or another system that monitors it would immediately detect the compromise. In essence, this means that the system fails "safe" so long as:
1. you have 2FA on your registry account, protecting privileged settings, etc
1. you use an automation token to push packages.

An attacker who gets hold of an automation token can push packages but these would be immediately detected because of the change of source. In a nutshell, you don't really need staged builds if you upload provenance from a trusted builder.

Reply:
> An attacker who gets hold of an automation token can push packages but these would be immediately detected because of the change of source. In a nutshell, you don't need staged builds if you upload provenance from a trusted builder.

💯 Yes great summary! We can also make it difficult to change the provenance, e.g. locking it to a particular source repo once set, or preventing publishing new versions that don't include it if it's previously been set. Changing this could be gated behind 2FA on your registry account, limited to admins for an org.

Reply:
If you disaggregate the system pushing the package to the repo from the system building/signing the artifact, you will be able to protect against the upload token being compromised because you can validate the signed artifacts with published public keys.

Furthermore, sigstore (I believe mentioned in this RFC, I'm still reading through it) can provide some ability to detect a compromised signing key. Will comment more on this later after reading more.